File: typescript/single-linked-list/list.ts
export interface LNode<T> {
    data: T;
    next: LNode<T> | null;
}

export interface LinkedList<T> {
    insertAtBegin(data: T): void;
    insertAtEnd(data: T): void;
    deleteNode(data: T): void;
    searchNode(data: T): boolean;
    printList(): void;
}

export class LNode<T> implements LNode<T> {
    data: T;
    next: LNode<T> | null;

    constructor(data: T) {
        this.data = data;
        this.next = null;
    }
}

export class LinkedList<T> implements LinkedList<T> {
    private head: LNode<T> | null;
    private size: number;

    constructor() {
        this.head = null;
        this.size = 0;
    }

    insertAtBegin(data: T): void {
        const newNode = new LNode(data);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }

    insertAtEnd(data: T): void {
        const newNode = new LNode(data);

        if (!this.head) {
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }
        this.size++;
    }

    deleteNode(data: T): void {
        let current = this.head;
        let previous: LNode<T> | null = null;

        while (current !== null) {
            if (current.data === data) {
                if (previous === null) {
                    this.head = current.next;
                } else {
                    previous.next = current.next;
                }
                this.size--;
                return;
            }
            previous = current;
            current = current.next;
        }
    }

    searchNode(data: T): boolean {
        let current = this.head;
        while (current !== null) {
            if (current.data === data) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    printList(): void {
        let current = this.head;
        let str = "";
        while (current) {
            str += current.data + " ";
            current = current.next;
        }
        console.log(str);
    }
}

File: rust/single-linked-list/src/list.rs
#[derive(Debug)]
struct Node {
    data: i32,
    next: Option<Box<Node>>,
}

pub struct LinkedList {
    head: Option<Box<Node>>,
}

impl LinkedList {
    pub fn new() -> Self {
        LinkedList { head: None }
    }

    pub fn insert_at_begin(&mut self, data: i32) {
        let new_node = Node {
            data,
            next: self.head.take(),
        };
        self.head = Some(Box::new(new_node));
    }

    pub fn insert_at_end(&mut self, data: i32) {
        let new_node = Box::new(Node {
            data,
            next: None,
        });

        match &mut self.head {
            None => {
                self.head = Some(new_node);
            }
            Some(head) => {
                let mut current = head;
                while current.next.is_some() {
                    current = current.next.as_mut().unwrap();
                }
                current.next = Some(new_node);
            }
        }
    }

    pub fn delete_node(&mut self, data: i32) {
        match &mut self.head {
            None => return,
            Some(head) => {
                if head.data == data {
                    self.head = head.next.take();
                    return;
                }

                let mut current = head;
                while current.next.is_some() {
                    if current.next.as_ref().unwrap().data == data {
                        current.next = current.next.take().unwrap().next;
                        return;
                    }
                    current = current.next.as_mut().unwrap();
                }
            }
        }
    }

    pub fn search_node(&self, data: i32) -> bool {
        let mut current = &self.head;
        while let Some(node) = current {
            if node.data == data {
                return true;
            }
            current = &node.next;
        }
        false
    }

    pub fn print_list(&self) {
        let mut current = &self.head;
        while let Some(node) = current {
            print!("{} ", node.data);
            current = &node.next;
        }
        println!();
    }
}